meetings/working-groups/collection-literals/Core-interface-target-type-proposal.md

```diff
@@ -0,0 +1,247 @@
+
+1. `new T[] { ... }`
+1. `new[] { ... }`
+1. `stackalloc T[] { ... }`
+1. `stackalloc[] { ... }`
+1. `new CollectionType<Etc> { ... }`
+1. `ImmutableCollectionType.Create(...)`
+1. `ImmutableCollection.CreateBuilder(); builder.Add(...); ...; builder.ToImmutable()`
+1. And more... including things like: `ImmutableArray<T>.Empty.Add(x).AddRange(y).etc()`
+
+An examination of the BCL and the top 20 Nuget packages (Newtonsoft, EF, Azure, Castle, AWS, Automapper, and more), all of which are >400m downloads, reveals very relevant data here.  Methods taking those interface collections account for roughly 28% of all methods taking some collection type (arrays, spans, other BCL collections), and `IEnumerable<T>` alone accounts for 25% of the collection-taking methods.  This is not surprising as our own practices, and general design guidance we give the community, is simply that methods should be extremely permissive in what they accept, and be precise in what they return.  `IEnumerable<T>` (and our other collection interfaces) act as that permissive type that we and our ecosystem have broadly adopted.
+
+Indeed, if it were not for `params T[]` (a full 50% of all collection-taking methods), `IEnumerable<T>` would be the most commonly taken collection by far for the ecosystem.
+
+Ideally, we would ship with support for everything, but we've currently made judicious moves from C#12 to C#13 based on complexity, but also based on impact.  For example, `Dictionary expressions` were moved to C#13 to lighten our load, and because data indicates that APIs that consume those dictionary types are only <3% of all apis that take collections in the first place.  `Natural type` support has also been pushed out because the complexity is felt to be substantive enough to warrant more time.
+
+With how important these interface types are though, we do not believe pushing out from C# 12 will allow us to ship a viable and coherent story to customers.  
+
+### What factors are at play?
+
+In any interesting design, there are many factors that must be assessed and considered as a holistic whole.  For collection expressions, the broad areas we want the land the feature are include, but are not limited to:
+
+1. Simplicity. Ideally the feature works in a fashion that is both simple to explain and simple for users to understand. Using some of our modern terminology, we'd like to avoid 'decoder rings' when people use it.
+
+2. Universality.  This restates the background context.  Namely that we want to meet the literal 97% case at launch.  This means needing good stories for arrays, spans, BCL concrete collections *and* BCL core interfaces.
```

Comment:
People may wonder... 97% case, as opposed to what percent if interfaces aren't allowed as targets? 95%? 50%? Can we ballpark a number for this?

Reply:
Yes.  If we take interfaces out, we go down to hitting 69% of all apis.  Importantly though, these APIs are modal.  So it's not like you're limited from calling the interface form, but then have another form (like the array/span) that is available to you.  Instead, it is just hte case that your only option is to fall back to explicitly instantiating a concrete type.

Also, The 97% comes from the remaining 3% being things in the dictionary family.  While present is reasonable numbers, it is dwarfed by the number of sequence-like collections everywhere.  This can also help explain why we felt ok (though not great) punting on dictionaries for 12.  It only limited you that much, as opposed to dropping 28% in the case of not supporting interfaces.

Reply:
```suggestion
2. Universality.  This restates the background context: we want to meet the literal 97% case at launch, not 69% which would be the case if interfaces can't be targeted.  This means needing good stories for arrays, spans, BCL concrete collections *and* BCL core interfaces.
```